File: MU/MU_3SL_interface_3.4.1/examples/calibration_online/calibration_online.c
/*
 * Software and its documentation is provided by iC-Haus GmbH or contributors "AS IS" and is
 * subject to the ZVEI General Conditions for the Supply of Products and Services with iC-Haus
 * amendments and the ZVEI Software clause with iC-Haus amendments (http://www.ichaus.de/EULA).
 */

#include "csv_file.h"
#include "mu_3sl_calibration_adjustments.h"
#include "mu_3sl_check_error_return.h"
#include "MU_3SL_defs.h"
#include "MU_3SL_interface.h"
#include "mu_3sl_nonius_curve.h"

#include <inttypes.h>
#include <math.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>


#if defined(_WIN32) || defined(__WIN32__)
    #include <windows.h>
#elif defined(linux) || defined(__linux) || defined(__APPLE__)
    #include <unistd.h>
#endif


inline void sleepMs(unsigned int ms)
{
#if defined(_WIN32) || defined(__WIN32__)
    Sleep(ms);
#elif defined(linux) || defined(__linux) || defined(__APPLE__)
    usleep(ms * 1000);
#endif
}


int main(int argc, char** argv)
{
    const double acquireFrameCycleTime_s  = 187.5E-6;
    const double acquireClockFrequency_hz = 2E6;
    // Calculation of the number of samples based on acquisition time in seconds
    const double acquisitionTime_s      = 5.0;
    const unsigned long numberOfSamples = lround(acquisitionTime_s / acquireFrameCycleTime_s);
    // const unsigned long numberOfSamples = 10000; // min. 16 (recommended 128) values per period

    const size_t maximumNumberOfAnalogCalibrationRuns             = 3;
    const double permissibleResidualErrorsDuringAnalogCalibration = 1.0;

    const char* noniusCurveCsvFilePath = argc <= 1 ? "nonius_curve.csv" : argv[1];

    const char* calibrationDataExportPath = "";

    uint16_t* masterRawData = malloc(numberOfSamples * sizeof(uint16_t));
    uint16_t* noniusRawData = malloc(numberOfSamples * sizeof(uint16_t));

    // Get the version of the library
    printf("Library version: %s%s\n\n", MU_GetVersionString(), MU_GetVersionSuffixString());

    // Construct a new instance of a virtual chip object and obtain a handle for it.
    MU_Handle muHandle;
    CHECK(MU_Open(&muHandle));

    // Connect the interface (USB-BiSS adapter) MB5U
    CHECK(MU_SetInterface(muHandle, MU_MB5U, ""));

    uint8_t revisionCode = MU_REV_MU128_X1;
    CHECK(MU_ReadChipRevision(muHandle, &revisionCode));
    CHECK(MU_UseRevision(muHandle, revisionCode));

    // Read all configuration parameters from the connected iC-MU* in the virtual chip object
    CHECK(MU_ReadParams(muHandle));

    MU_Calibration* calibration = MU_getCalibration(muHandle);

    uint32_t muMPC;
    CHECK(MU_GetParam(muHandle, MU_MPC, NULL, &muMPC));
    uint8_t masterPeriodCode = (uint8_t)muMPC;


    printf("---------------------------------------------------------------\n"
           "- Reset current analog track adjustments (loaded from EEPROM) -\n"
           "---------------------------------------------------------------\n");

    MU_Calibration_AnalogTrackAdjustments initialMasterAdjustments = {0, 0, 0, 0, 0};
    MU_Calibration_AnalogTrackAdjustments initialNoniusAdjustments = {0, 0, 0, 0, 0};
    MU_Calibration_setCurrentAnalogTrackAdjustments(
            calibration, &initialMasterAdjustments, &initialNoniusAdjustments);
    CHECK(MU_setCalibration(muHandle, calibration));
    CHECK(MU_WriteParams(muHandle, false, NULL));


    printf("Initial iC-MU analog parameters:\n");
    printAnalogAdjustments(calibration);
    printf("\n\n");

    printf("---------------------------------------------------------------\n"
           "------------- Activate calibration configuration --------------\n"
           "---------------------------------------------------------------\n"
           "\n");
    CHECK(MU_activateCalibrationConfig(muHandle));

    MU_CalibrationAnalyzeResult* analyzeResult = NULL;

    for (size_t i = 0; i < maximumNumberOfAnalogCalibrationRuns; ++i) {
        printf("---------------------------------------------------------------\n"
               "---------------------- Acquire raw data -----------------------\n"
               "---------------------------------------------------------------\n"
               "\n");
        if (MU_acquireRawData(
                    muHandle,
                    masterRawData,
                    noniusRawData,
                    numberOfSamples,
                    0,
                    acquireFrameCycleTime_s,
                    acquireClockFrequency_hz)) {
            MU_Error lastError;
            MU_ErrorType errorType;
            char errorText[1024];
            MU_GetLastError(muHandle, &lastError, &errorType, errorText);
            fprintf(stderr, "MU_acquireRawData error: %s\n", errorText);

            MU_deactivateCalibrationConfig(muHandle);
            free(masterRawData);
            free(noniusRawData);
            MU_Calibration_delete(calibration);
            MU_Close(muHandle);
            exit(EXIT_FAILURE);
        }

        char calibrationDataFilePath[256];
        snprintf(
                calibrationDataFilePath,
                sizeof(calibrationDataFilePath) / sizeof(char),
                "%scalibration_data_%" PRIuPTR ".csv",
                calibrationDataExportPath,
                i);
        MU_Calibration_AnalogTrackAdjustments masterTrackAdjustments;
        MU_Calibration_getAnalogMasterTrackAdjustments(calibration, &masterTrackAdjustments);
        MU_Calibration_AnalogTrackAdjustments noniusTrackAdjustments;
        MU_Calibration_getAnalogNoniusTrackAdjustments(calibration, &noniusTrackAdjustments);
        writeCalibrationDataCSVFile(
                masterRawData,
                noniusRawData,
                numberOfSamples,
                calibrationDataFilePath,
                &revisionCode,
                &masterPeriodCode,
                &masterTrackAdjustments,
                &noniusTrackAdjustments);

        printf("---------------------------------------------------------------\n"
               "---------------------- Analyze raw data  ----------------------\n"
               "---------------------------------------------------------------\n");
        if (analyzeResult != NULL) {
            MU_CalibrationAnalyzeResult_delete(analyzeResult);
        }
        analyzeResult = MU_Calibration_analyzeRawData(
                calibration, masterRawData, noniusRawData, numberOfSamples);
        if (analyzeResult == NULL) {
            fprintf(stderr, "Raw data are not analyzable\n");
            MU_deactivateCalibrationConfig(muHandle);
            free(masterRawData);
            free(noniusRawData);
            MU_Calibration_delete(calibration);
            MU_Close(muHandle);
            exit(EXIT_FAILURE);
        }

        printAnalyzeResultLog(analyzeResult);

        printf("Residual errors (relative changes in \"LSB\")\n");
        printRelativeAdjustments(analyzeResult);

        printAnalogAnalyzeResultAdjustableLog(calibration, analyzeResult);

        MU_Calibration_RelativeAnalogTrackAdjustments relativeMasterTrackAdjustments;
        MU_Calibration_getRelativeMasterTrackAdjustments(
                analyzeResult, &relativeMasterTrackAdjustments);
        MU_Calibration_RelativeAnalogTrackAdjustments relativeNoniusTrackAdjustments;
        MU_Calibration_getRelativeNoniusTrackAdjustments(
                analyzeResult, &relativeNoniusTrackAdjustments);

        if (fabs(relativeMasterTrackAdjustments.cosineGain_lsb)
                    <= permissibleResidualErrorsDuringAnalogCalibration
            && fabs(relativeNoniusTrackAdjustments.cosineGain_lsb)
                       <= permissibleResidualErrorsDuringAnalogCalibration
            && fabs(relativeMasterTrackAdjustments.sineOffset_lsb)
                       <= permissibleResidualErrorsDuringAnalogCalibration
            && fabs(relativeNoniusTrackAdjustments.sineOffset_lsb)
                       <= permissibleResidualErrorsDuringAnalogCalibration
            && fabs(relativeMasterTrackAdjustments.cosineOffset_lsb)
                       <= permissibleResidualErrorsDuringAnalogCalibration
            && fabs(relativeNoniusTrackAdjustments.cosineOffset_lsb)
                       <= permissibleResidualErrorsDuringAnalogCalibration
            && fabs(relativeMasterTrackAdjustments.phase_lsb)
                       <= permissibleResidualErrorsDuringAnalogCalibration
            && fabs(relativeNoniusTrackAdjustments.phase_lsb)
                       <= permissibleResidualErrorsDuringAnalogCalibration) {
            printf("\n"
                   "All residual errors (absolute relative changes in \"LSB\") are smaller than "
                   "%.3f.\n"
                   "From this point, no further analog calibration step would be required.\n"
                   "\n",
                   permissibleResidualErrorsDuringAnalogCalibration);
            break;
        }

        printf("---------------------------------------------------------------\n"
               "------------ Adjust the analyzed analog parameters ------------\n"
               "---------------------------------------------------------------\n");

        MU_Calibration_adjustAnalogByAnalyzeResult(calibration, analyzeResult);

        printf("iC-MU analog parameters after adjustment:\n");
        printAnalogAdjustments(calibration);
        printf("\n\n");

        MU_setCalibration(muHandle, calibration);
        CHECK(MU_WriteParams(muHandle, false, NULL));
    }

    printf("\n"
           "---------------------------------------------------------------\n"
           "------------ Adjust the analyzed nonius parameters ------------\n"
           "---------------------------------------------------------------\n"
           "\n");

    MU_Calibration_NoniusTrackOffsetTable optimizedNoniusTrackOffsetTable;
    MU_Calibration_getOptimizedNoniusTrackOffsetTable(
            analyzeResult, &optimizedNoniusTrackOffsetTable);
    MU_Calibration_setCurrentNoniusTrackOffsetTable(calibration, &optimizedNoniusTrackOffsetTable);

    MU_CalibrationAnalyzeResult_delete(analyzeResult);
    // Generate a new analysis with the optimized nonius track offset table
    analyzeResult = MU_Calibration_analyzeRawData(
            calibration, masterRawData, noniusRawData, numberOfSamples);

    optionalPrintOptimizedNoniusTrackOffsetTable(analyzeResult, noniusCurveCsvFilePath);
    printf("\n");
    optionalPrintOptimizedNoniusTrackOffsetParameters(analyzeResult);

    MU_setCalibration(muHandle, calibration);
    CHECK(MU_WriteParams(muHandle, false, NULL));

    MU_CalibrationAnalyzeResult_delete(analyzeResult);
    MU_Calibration_delete(calibration);

    printf("\n"
           "---------------------------------------------------------------\n"
           "------------ Deactivate calibration configuration -------------\n"
           "---------------------------------------------------------------\n");
    CHECK(MU_deactivateCalibrationConfig(muHandle));

    printf("\n"
           "---------------------------------------------------------------\n"
           "------------- iC-MU CRC calculation and reset ABS -------------\n"
           "---------------------------------------------------------------\n");

    CHECK(MU_WriteCmdRegister(muHandle, MU_CMD_CRC_CALC));
    CHECK(MU_WriteCmdRegister(muHandle, MU_CMD_ABS_RESET));

    printf("\n"
           "---------------------------------------------------------------\n"
           "---------- Save the configuration in the chip EEPROM ----------\n"
           "---------------------------------------------------------------\n");

    CHECK(MU_WriteCmdRegister(muHandle, MU_CMD_WRITE_ALL));

    free(masterRawData);
    free(noniusRawData);
    MU_Close(muHandle);

    return EXIT_SUCCESS;
}

File: MU/MU_3SL_interface_3.4.1/wrapper/python/examples/calibration_online.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Copyright 2022 iC-Haus GmbH

Software and its documentation is provided by iC-Haus GmbH or contributors "AS IS" and is
subject to the ZVEI General Conditions for the Supply of Products and Services with iC-Haus
amendments and the ZVEI Software clause with iC-Haus amendments (http://www.ichaus.de/EULA).
"""

import mu_3sl_interface as mu_3sl


acquire_frame_cycle_time_s = 187.5E-6
acquire_clock_frequency_hz = 2E6
# Calculation of the number of samples based on acquisition time in seconds
acquisition_time_s = 5.0
number_of_samples = round(acquisition_time_s / acquire_frame_cycle_time_s)
# number_of_samples = 10000; # min. 16 (recommended 128) values per period

print("Library version: {}".format(mu_3sl.__version__))

# Construct a new instance of a virtual chip object and obtain a handle for it.
mu = mu_3sl.MuHandle()

# Connect the interface (USB-BiSS adapter) MB5U
mu.set_interface(mu_3sl.Interface.MB5U, "")

revisionCode = mu.read_chip_revision()
mu.use_revision(revisionCode)

# Read all configuration parameters from the connected iC-MU* in the virtual chip object
mu.read_parameters()

calibration = mu.get_calibration()

master_period_code = mu.get_parameter(mu_3sl.Parameter.MPC)

print("\n" +
      "---------------------------------------------------------------\n" +
      "- Reset current analog track adjustments (loaded from EEPROM) -\n" +
      "---------------------------------------------------------------")

initialMasterAdjustments = mu_3sl.AnalogTrackAdjustments()
initialMasterAdjustments.cosine_gain = 0
initialMasterAdjustments.sine_offset = 0
initialMasterAdjustments.cosine_offset = 0
initialMasterAdjustments.phase = 0
initialMasterAdjustments.phase_range = 0

initialNoniusAdjustments = mu_3sl.AnalogTrackAdjustments()
initialNoniusAdjustments.cosine_gain = 0
initialNoniusAdjustments.sine_offset = 0
initialNoniusAdjustments.cosine_offset = 0
initialNoniusAdjustments.phase = 0
initialNoniusAdjustments.phase_range = 0

calibration.set_current_analog_track_adjustments(initialMasterAdjustments, initialNoniusAdjustments)

mu.set_calibration(calibration)
mu.write_parameters()

print("Initial iC-MU signal conditioning parameters:")


def print_analog_track_adjustments(adjustments: mu_3sl.AnalogTrackAdjustments,
                                   track_short_name: str):
    print(("Cosine gain (GX_{0}):     {1:4d} (0x{1:04X})\n" +
           "Sine offset (VOSS_{0}):   {2:4d} (0x{2:04X})\n" +
           "Cosine offset (VOSC_{0}): {3:4d} (0x{3:04X})\n" +
           "Phase (PH_{0}):           {4:4d} (0x{4:04X})\n" +
           "Phase range (PHR_{0}):    {5:4d} (0x{5:04X})").format(track_short_name,
                                                                  adjustments.cosine_gain,
                                                                  adjustments.sine_offset,
                                                                  adjustments.cosine_offset,
                                                                  adjustments.phase,
                                                                  adjustments.phase_range))


def print_analog_adjustments(calibration: mu_3sl.Calibration):
    masterTrackAdjustments = calibration.analog_master_track_adjustments()
    print_analog_track_adjustments(masterTrackAdjustments, "M")
    noniusTrackAdjustments = calibration.analog_nonius_track_adjustments()
    print_analog_track_adjustments(noniusTrackAdjustments, "N")


print_analog_adjustments(calibration)
print("")

print("\n" +
      "---------------------------------------------------------------\n" +
      "----------------- Analog Calibration of iC-MU -----------------\n" +
      "---------------------------------------------------------------")

mu.activate_calibration_configuration()
try:
    masterRawData, noniusRawData = mu.acquire_raw_data(number_of_samples, 0,
                                                       acquire_frame_cycle_time_s,
                                                       acquire_clock_frequency_hz)
finally:
    mu.deactivate_calibration_configuration()

analyze_result = calibration.analyze_raw_data(masterRawData, noniusRawData)

print(analyze_result.log(mu_3sl.CalibrationAnalyzeResultLogLevel.ALL))

print("Relative track adjustments (relative changes in \"LSB\")")


def print_relative_adjustments(analyze_result: mu_3sl.AnalyzeResult):
    relativeMasterTrackAdjustments = analyze_result.relative_master_track_adjustments()
    relativeNoniusTrackAdjustments = analyze_result.relative_nonius_track_adjustments()
    print(("Track:             Master |   Nonius\n" +
           "  Cosine gain:   {:8.4f} | {:8.4f}\n" +
           "  Sine offset:   {:8.4f} | {:8.4f}\n" +
           "  Cosine offset: {:8.4f} | {:8.4f}\n" +
           "  Phase adjust:  {:8.4f} | {:8.4f}\n").format(
        relativeMasterTrackAdjustments.cosine_gain_lsb,
        relativeNoniusTrackAdjustments.cosine_gain_lsb,
        relativeMasterTrackAdjustments.sine_offset_lsb,
        relativeNoniusTrackAdjustments.sine_offset_lsb,
        relativeMasterTrackAdjustments.cosine_offset_lsb,
        relativeNoniusTrackAdjustments.cosine_offset_lsb,
        relativeMasterTrackAdjustments.phase_lsb,
        relativeNoniusTrackAdjustments.phase_lsb)
    )


print_relative_adjustments(analyze_result)

is_adjustable, adjustment_message = calibration.is_analog_analyze_result_adjustable(
    analyze_result,
    mu_3sl.CalibrationAnalyzeResultLogLevel.ALL)
if is_adjustable:
    print("Analyze result is adjustable.\nMessage:\n{}".format(adjustment_message))
else:
    print("Analyze result is not adjustable.\nMessage:\n{}".format(adjustment_message))

calibration.adjust_analog_by_analyze_result(analyze_result)

print("iC-MU signal conditioning parameters after calibration:")
print_analog_adjustments(calibration)

mu.set_calibration(calibration)
mu.write_parameters()

print("\n" +
      "---------------------------------------------------------------\n" +
      "----------------- Nonius Calibration of iC-MU -----------------\n" +
      "---------------------------------------------------------------")

mu.activate_calibration_configuration()
try:
    masterRawData, noniusRawData = mu.acquire_raw_data(number_of_samples, 0,
                                                       acquire_frame_cycle_time_s,
                                                       acquire_clock_frequency_hz)
finally:
    mu.deactivate_calibration_configuration()

analyze_result = calibration.analyze_raw_data(masterRawData, noniusRawData)

print(analyze_result.log(mu_3sl.CalibrationAnalyzeResultLogLevel.ALL))

print("Residual errors (relative changes in \"LSB\")")
print_relative_adjustments(analyze_result)



def optionalPrintOptimizedNoniusTrackOffsetTable(analyze_result: mu_3sl.AnalyzeResult,
                                                 master_period_code: int):
    nonius_track_offset_table = analyze_result.optimized_nonius_track_offset_table()
    print("Optimized nonius track offset table:")
    print("SPO_BASE: {:2d}".format(nonius_track_offset_table.spo_base))
    for i in range(16):
        if i >= 10:
            print("SPO_{:d}:   {:2d}".format(i, nonius_track_offset_table[i]))
        else:
            print("SPO_{:d}:    {:2d}".format(i, nonius_track_offset_table[i]))
    print("")

    nonius_calibration_max_allow_phase_error = 1 << (14 - 1 - master_period_code)
    nonius_phase_margin_max = analyze_result.nonius_phase_margin_max()
    nonius_phase_margin_min = analyze_result.nonius_phase_margin_min()
    nonius_phase_margin_max_percent = (
            nonius_phase_margin_max / nonius_calibration_max_allow_phase_error * 100)
    nonius_phase_margin_min_percent = (
            nonius_phase_margin_min / -nonius_calibration_max_allow_phase_error * 100)
    print("Nonius phase margin: Max.: {:d} ({:.3f}%) Min.: {:d} ({:.3f}%)".format(
        nonius_phase_margin_max,
        nonius_phase_margin_max_percent,
        nonius_phase_margin_min,
        nonius_phase_margin_min_percent)
    )

    print("Nonius phase range limit: {:d}".format(analyze_result.nonius_phase_range_limit()))

    print("Nonius upper phase margin: {:f}%", analyze_result.nonius_upper_phase_margin())
    print("Nonius lower phase margin: {:f}%", analyze_result.nonius_lower_phase_margin())


optionalPrintOptimizedNoniusTrackOffsetTable(analyze_result, master_period_code)

optimized_nonius_track_offset_table = analyze_result.optimized_nonius_track_offset_table()
calibration.set_current_nonius_track_offset_table(optimized_nonius_track_offset_table)

mu.set_calibration(calibration)
mu.write_parameters()

mu.write_cmd_register(mu_3sl.Command.CRC_CALC)
mu.write_cmd_register(mu_3sl.Command.ABS_RESET)

print("\n" +
      "---------------------------------------------------------------\n" +
      "--------------- Send WRITE_ALL Command to iC-MU ---------------\n" +
      "---------------------------------------------------------------")

mu.write_cmd_register(mu_3sl.Command.WRITE_ALL)
